nixos/modules/services/misc/forgejo.nix

```diff
@@ -590,15 +604,16 @@ in
         ReadWritePaths = [ cfg.customDir cfg.dump.backupDir cfg.repositoryRoot cfg.stateDir cfg.lfs.contentDir ];
         UMask = "0027";
         # Capabilities
-        CapabilityBoundingSet = "";
+        AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" ];
+        CapabilityBoundingSet = [ "CAP_NET_BIND_SERVICE" ];
```

Comment:
Should we maybe only  get this, when `port < 1024`?

Often deployments will be behind nginx and on unprivileged ports.

Reply:
This was discussed previously, see https://github.com/NixOS/nixpkgs/pull/306457#discussion_r1592984395. I think it's a reasonable change, and from reading the documention on capabilities(https://www.man7.org/linux/man-pages/man7/capabilities.7.html) I don't see any risk in giving this permission, as the most that a malicious actor could potentially do with it is bind to a low port. I don't see this as a security issue to have with ports > 1024.

Reply:
> This was discussed previously

Oh, sorry, I was out of the loop for that.

> What do you think

Note that I'm not necessarily against just granting the CAP, as previously asked by @emilylange. I agree that the security implications are minimal (though not zero), since CI runners are part of a different service def.

What I don't like is tacking a change like this on a PR of unrelated description. But since all the current maintainers are here now, I won't block it.

I like the separate option. It has the advantage of being explicit over an automatic detection of `port < 1024` and breaking a security boundary should always be as explicit as conveniently possible. Maybe `services.forgejo.allowBindPrivilegedPort`?

Reply:
With the option, we could even implement a helpful error message, to set the option if the bind port is privileged.

Reply:
> > This was discussed previously
> 
> Oh, sorry, I was out of the loop for that.
> 
> > What do you think
> 
> Note that I'm not necessarily against just granting the CAP, as previously asked by @emilylange. I agree that the security implications are minimal (though not zero), since CI runners are part of a different service def.
> 
> What I don't like is tacking a change like this on a PR of unrelated description. But since all the current maintainers are here now, I won't block it.
> 
> I like the separate option. It has the advantage of being explicit over an automatic detection of `port < 1024` and breaking a security boundary should always be as explicit as conveniently possible. Maybe `services.forgejo.allowBindPrivilegedPort`?

So again, as mentioned, I think that a separate option would be good, but we would need to always make sure that we check all options taking a port to ensure all of them are above 1024 before surfacing this warning. As also discussed, this is a very small security decrease, and going from my experience, it doesn't really change the attack surface much at all. Therefore, I think unconditionally enabling this is fine, especially if you are running the service behind a reverse proxy, as is generally suggested.

Reply:
Yeah, so as I said, I was out of the loop for the initial round of feedback and I'm fine going with what @emilylange already asked of you and I don't think this is worth endlessly nitpicking. Since I already expressed my preference (the capability change in a separate PR), but indicated that I won't block, I'll resolve this thread now.

For posterity, I'll leave the lengthier response in, since I've already typed it out. Beware: Severe nitpicking ahead

---

> very small security decrease

The reason I've flagged this, is less about the absolute amount of security change (in fact, I think linux could do without privileged network ports), but more about default expectations when dealing with a linux system. For example, I could easily see a firewall accepting all privileged ports by default, which could easily "amplify" the small attack surface.

But yeah, much of this is hypothetical, because the capability would only be given to forgejo, and not the CI runner, which seems much more vulnerable than the main forgejo process.

Still, as I said before, the thing that bugs me here is PR hygiene: The title here is still "Cleanup environment settings", where it's not just a cleanup, but also a capability change.

> Therefore, I think unconditionally enabling this is fine, especially if you are running the service behind a reverse proxy

I'm having a little trouble tracking this implication, sorry: Are we talking about a reverse proxy on a different machine?

Even when running a server without a reverse proxy, I prefer binding to high ports and then just redirecting the appropriate low ports via firewall.

And with a reverse proxy (on the same machine), I try to not bind network ports at all, but `+unix` domain sockets, because those can also resist local attackers.

Reply:
I think it's a fair nit that the PR title wasn't appropriate, so I've updated it. Thanks for suggesting that, since the old title definitely didn't include the full scope of this PR.